runner/startup.sh

```diff
@@ -1,70 +1,75 @@
+    sudo tee -a $dockerd_supervisor_config_path 1>&- <<<"environment=DOCKERD_ROOTLESS_ROOTLESSKIT_MTU=$MTU"
+    dockerd_config=$(jq -S ".mtu = $MTU" <<<"$dockerd_config")
+  else
+    # shellcheck source=runner/logger.bash
+    source logger.bash
+    log.error "Docker MTU must be an integer, continuing bootstrapping without it, got: $MTU"
+    MTU=
+  fi
 fi
 
-if [ -n "${DOCKER_REGISTRY_MIRROR}" ]; then
-jq ".\"registry-mirrors\"[0] = \"${DOCKER_REGISTRY_MIRROR}\"" /etc/docker/daemon.json > /tmp/.daemon.json && mv /tmp/.daemon.json /etc/docker/daemon.json
+if [[ ${DOCKER_REGISTRY_MIRROR:-} != '' ]]; then
+  dockerd_config=$(jq -S --arg url "$DOCKER_REGISTRY_MIRROR" '."registry-mirrors" += ["$url"]' <<<"$dockerd_config")
 fi
-SCRIPT
 
-dump() {
-  local path=${1:?missing required <path> argument}
-  shift
-  printf -- "%s\n---\n" "${*//\{path\}/"$path"}" 1>&2
-  cat "$path" 1>&2
-  printf -- '---\n' 1>&2
-}
+sudo mkdir -p ${dockerd_config_path%/*}
+sudo tee $dockerd_config_path 1>&- <<<"$dockerd_config"
 
-for config in /etc/docker/daemon.json /etc/supervisor/conf.d/dockerd.conf; do
-  dump "$config" 'Using {path} with the following content:'
+for config in $dockerd_config_path $dockerd_supervisor_config_path; do
+  (
+    echo "Using '$config' with the following content <<CONFIG"
+    cat $config
+    echo 'CONFIG'
+  ) 1>&2
 done
 
-log.debug 'Starting supervisor daemon'
-sudo /usr/bin/supervisord -n >> /dev/null 2>&1 &
-
-log.debug 'Waiting for processes to be running...'
-processes=(dockerd)
-
-for process in "${processes[@]}"; do
-    wait_for_process "$process"
-    if [ $? -ne 0 ]; then
-        log.error "$process is not running after max time"
-        dump /var/log/dockerd.err.log 'Dumping {path} to aid investigation'
-        exit 1
-    else
-        log.debug "$process is running"
-    fi
-done
+sudo /usr/bin/supervisord -c /etc/supervicor/supervisord.conf
 
-if [ -n "${MTU}" ]; then
-  sudo ifconfig docker0 mtu ${MTU} up
+if [[ $MTU ]]; then
+  # It might take a few ticks for the `docker0` device to be up after starting
+  # the docker daemon with supervisord, hence, we retry for a little while.
+  # Notice how we use `nice` to execute the command with a least favorable
+  # scheduling priority and are using a noop (`:`) instruction instead of
+  # `sleep`. We do this to ensure that we keep the startup as fast as possible.
+  if ! nice -n 19 timeout -k 70 60 -- bash -c "(until sudo ifconfig docker0 mtu '$MTU' up; do :; done) 1>&- 2>&-"; then
+    # shellcheck source=runner/logger.bash
+    source logger.bash
+    log.error 'Failed to set docker interface mtu within 1 minute, continuing bootstrapping without it...'
+  fi
 fi
 
-# Wait processes to be running
-entrypoint.sh
+# We leave the wait for docker to come up to our entrypoint.sh script, since its
+# check properly handles the case where the docker service is in a crash loop.
+# Users still have the ability to disable the wait.
+export DOCKER_ENABLED=true
```

Comment:
why are we setting this in the startup? The controller is adding this env to the runner container regardless

Reply:
It could be set to `DOCKER_ENABLED=false` which would be incorrect in this case since we just started it and know that it is enabled. We want the `entrypoint.sh` script to verify that it is running before starting the run itself.

That said, I have to admit that I do not fully understand why we have two variables for the seemingly same thing in our `entrypoint.sh` script, but maybe you know more and can tell, since that information might change the implementation here.